### PR TITLE
Move matplotlib code into to the backend

### PR DESCRIPTION
**Pull request by swright87**
# Related issues

N/A

# Proposed changes

This PR moves the matplotlib code out of the `cascade` and `navchart` upper level files and into the `matplotlib` backend.

This makes it clearer to extend the library with additional backends.

### CODE REVIEW

File: p3/plot/_cascade.py [hunk 1]
```diff
@@ -6,225 +6,8 @@
...
-    elif isinstance(kwarg, matplotlib.colors.Colormap):
-        cmap = kwarg
-    else:
-        raise ValueError("Unsupported type provided for colormap.")
-
-    cmap = cmap.resampled(len(applications))
-    colors = cmap(np.linspace(0, 1, len(applications)))
-    return {app: color for app, color in zip(applications, colors)}
+from p3._utils import _require_columns
+from p3.plot.backend.matplotlib import CascadePlot as MPLCascadePlot
```
Comment (Pennycook) [inline p3/plot/_cascade.py:10]: Doing things this way means that matplotlib has to be installed even when the user is directing things to another backend.  With multiple backends, the user would have to install all the backends just to use one of them.

So we should remove this:
```suggestion
```

See below for my suggested fix.  Again, note that it should apply to the equivalent lines for navigation charts too.

File: p3/plot/_cascade.py [hunk 2]
```diff
@@ -301,133 +84,12 @@ def cascade(df, eff=None, size=(6, 5), **kwargs):
...
-    if len(efficiency_columns) == 0:
-        msg = "DataFrame does not contain an 'app eff' or 'arch eff' column."
-        raise ValueError(msg)
-
-    if eff is None:
-        eff_column = efficiency_columns[0]
+    kwargs.setdefault("backend", "matplotlib")
+    backend = kwargs["backend"]
+    if backend == "matplotlib":
+        return MPLCascadePlot(df, eff, size, **kwargs)
```
Comment (Pennycook) [inline p3/plot/_cascade.py:93]: We can defer the import until we need it like this:

```suggestion
        from p3.plot.backend.matplotlib import CascadePlot
        return CascadePlot(df, eff, size, **kwargs)
```

...and that way, there's no dependency on matplotlib unless it's the backend in use.

**Review by Pennycook — CHANGES_REQUESTED** [review]
Thank you for doing this.  The refactor is a lot cleaner, and it was much easier to review this without the PGF plots changes mixed in.

Things mostly look good to me.  I have a couple of small suggestions, and I'd like to hear what @laserkelvin thinks too.

File: p3/plot/backend/matplotlib.py [hunk 3]
```diff
@@ -49,7 +324,174 @@ def get_axes(self, subplot="eff"):
             return self.axes[0][1]
         if subplot == "plat":
             return self.axes[1][0]
-        raise ValueError("Unrecognized subplot name: '%s'" % subplot)
+        raise ValueError("Unrecognized subplot name: '%s'" % subplot)  # noqa
```
[the remark this answers is not in the document]
Comment (swright87) [reply p3/plot/backend/matplotlib.py:327]: It seems this issue was kicked up by the CI. I'm guessing it doesn't like the % in the string and a % for the substitution. Every other instance I can see seems to have the message in a `msg` string and then uses the substitution on a subsequent, so I've swapped to that style here (though perhaps fstrings are the 'pythonic' way?).

File: p3/plot/_cascade.py (hunk 1 above)
Comment (swright87) [reply p3/plot/_cascade.py:10]: Agree. Done.

File: p3/plot/_cascade.py (hunk 2 above)
Comment (swright87) [reply p3/plot/_cascade.py:93]: Done.

**Review by Pennycook — APPROVED** [review]
This looks good to me, but I'd like @laserkelvin to weigh in too before I merge anything.

@swright87: This is really pedantic, but I'd also prefer if your commit message didn't end with a period.  Would you mind changing that?

**Review by laserkelvin — APPROVED** [review]

**Comment by swright87** [comment]
> This looks good to me, but I'd like @laserkelvin to weigh in too before I merge anything.
> 
> @swright87: This is really pedantic, but I'd also prefer if your commit message didn't end with a period. Would you mind changing that?

Done (and fixed the previous commit message with a period too...).

**Pull request merged**